File: ruby/01_from_push_to_pull/lib/pushing_balances_calculator.rb
require_relative 'values_of_month'

class PushingBalancesCalculator
  attr_reader :transactions
  private :transactions
  def initialize(transactions)
    @transactions = transactions
  end

  def fill_data(balances_of_month_list)
    values_of_month = ValuesOfMonth.new
    balance = 0

    balances_of_month_list.each do |balances_of_month|
      ultimo = balances_of_month.date.day

      average_balance = 0.0
      day_of_latest_balance = 1
      transactions_of_month = transactions_of_month(balances_of_month.date)
      transactions_of_month.each do |transaction|
        day = transaction.date.day
        average_balance += calculate_proportional_balance(day_of_latest_balance, balance, day, ultimo)
        balance += transaction.amount
        day_of_latest_balance = day
      end

      average_balance += calculate_proportional_balance(day_of_latest_balance, balance, ultimo + 1, ultimo)
      values_of_month.set_balance_and_average(balance, average_balance)

      balances_of_month.balance = values_of_month.balance
      balances_of_month.average_balance = values_of_month.average_balance
    end
  end

  private
  def calculate_proportional_balance(day_of_latest_balance, balance, day, days_in_month)
    counting_days = day - day_of_latest_balance
    return 0 if counting_days == 0
    rate = counting_days.to_f / days_in_month.to_f
    balance * rate
  end

  def transactions_of_month(date)
    transactions.select { |transaction|
      are_same_month_and_year(date, transaction.date)
    }
  end

  def are_same_month_and_year(date, date_of_transaction)
    date_of_transaction.month == date.month &&
      date_of_transaction.year == date.year
  end
end

File: csharp/02_FromPushToPull/PushingBalancesCalculator.cs
using System;
using System.Collections.Generic;
using Common;

namespace FromPushToPull
{
    public class PushingBalancesCalculator : BalancesOfMonthCalculator
    {
        private readonly IList<Transaction> transactions;

        public PushingBalancesCalculator(IList<Transaction> transactions)
        {
            this.transactions = transactions;
        }

        public void FillData(IList<BalancesOfMonth> balancesOfMonthList)
        {
            ValuesOfMonth valuesOfMonth = new ValuesOfMonth();
            int balance = 0;

            foreach (BalancesOfMonth balancesOfMonth in balancesOfMonthList)
            {
                DateTime dateOfMonth = balancesOfMonth.Date;
                IList<Transaction> transactionsOfMonth = TransactionsOfMonth(dateOfMonth);

                int ultimo = dateOfMonth.Day;

                double averageBalance = 0;
                int dayOfLatestBalance = 1;
                foreach (Transaction transaction in transactionsOfMonth)
                {
                    int day = transaction.Date.Day;
                    averageBalance += CalculateProportionalBalance(dayOfLatestBalance, balance, day, ultimo);
                    balance += transaction.Amount;
                    dayOfLatestBalance = day;
                }

                averageBalance += CalculateProportionalBalance(dayOfLatestBalance, balance, ultimo + 1, ultimo);

                valuesOfMonth.Balance = balance;
                valuesOfMonth.AverageBalance = (int)averageBalance;

                balancesOfMonth.Balance = valuesOfMonth.Balance;
                balancesOfMonth.AverageBalance = valuesOfMonth.AverageBalance;
            }
        }

        private double CalculateProportionalBalance(int dayOfLatestBalance, int balance, int day, int daysInMonth)
        {
            int countingDays = day - dayOfLatestBalance;
            if (countingDays == 0)
            {
                return 0;
            }
            double rate = (double)countingDays / daysInMonth;
            return (balance * rate);
        }

        private IList<Transaction> TransactionsOfMonth(DateTime date)
        {
            IList<Transaction> results = new List<Transaction>();
            foreach (Transaction transaction in transactions)
            {
                DateTime dateOfTransaction = transaction.Date;
                if (AreSameMonthAndYear(date, dateOfTransaction))
                {
                    results.Add(transaction);
                }
            }
            return results;
        }

        private bool AreSameMonthAndYear(DateTime date, DateTime dateOfTransaction)
        {
            return dateOfTransaction.Month == date.Month && dateOfTransaction.Year == date.Year;
        }

    }
}
